[JSC] Math.random JIT path uses arithmetic instead of logical right shifts

matthewoestreich opened the pull request:
#### cf0e17e68442142b7024bcc21f4170a4c90c71fa
<pre>
[JSC] Math.random JIT path uses arithmetic instead of logical right shifts
<a href="https://bugs.webkit.org/show_bug.cgi?id=299242">https://bugs.webkit.org/show_bug.cgi?id=299242</a>

Reviewed by Justin Michaud.

--SUMMARY--
This report concerns Math.random output when taking the JIT path.
Currently, state advancement in emitRandomThunkImpl  uses arithmetic
right shifts (rshift64), whereas all other advancement paths use logical
right shifts.
[Source]: Source/JavaScriptCore/jit/AssemblyHelpers.cpp, line 847.

--REFERENCE IMPLEMENTATION--
WeakRandom::nextState uses the &gt;&gt; operator on a uint64_t, which performs
a logical shift.
[Source]: Source/WTF/wtf/WeakRandom.h, line 109.

--FTL LOWERING PATH--
compileArithRandom uses lShr (logical shift right).
[Source]: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp, line 3501.

--JIT PATH ((problem))--
emitRandomThunkImpl calls jit.rshift64, which emits SAR (arithmetic shift).
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5311.
By contrast, urshift64 emits SHR (logical shift), this matches the intended
semantics.
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5350.

--IMPACT--
This discrepancy means that, when given the same seed, Math.random produces
different output depending on whether the JIT path is taken. The bug is not
immediately obvious because both arithmetic and logical right shifts still
generate values within the expected [0, 1) range, and the distribution may look
reasonable at a glance. However, the underlying state advancement diverges from
the intended algorithm, which leads to inconsistent and incorrect behavior
compared to other code paths.

Test: JSTests/stress/math-random-inline-vs-non-inline-paths.js

Test: JSTests/stress/math-random-inline-vs-non-inline-paths.js
* JSTests/stress/math-random-inline-vs-non-inline-paths.js: Added.
(isEqualArr):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::emitRandomThunkImpl):

[JSC] Math.random JIT path uses arithmetic instead of logical right shifts
<a href="https://bugs.webkit.org/show_bug.cgi?id=299242">https://bugs.webkit.org/show_bug.cgi?id=299242</a>

Reviewed by Justin Michaud.

--SUMMARY--
This report concerns Math.random output when taking the JIT path.
Currently, state advancement in emitRandomThunkImpl  uses arithmetic
right shifts (rshift64), whereas all other advancement paths use logical
right shifts.
[Source]: Source/JavaScriptCore/jit/AssemblyHelpers.cpp, line 847.

--REFERENCE IMPLEMENTATION--
WeakRandom::nextState uses the &gt;&gt; operator on a uint64_t, which performs
a logical shift.
[Source]: Source/WTF/wtf/WeakRandom.h, line 109.

--FTL LOWERING PATH--
compileArithRandom uses lShr (logical shift right).
[Source]: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp, line 3501.

--JIT PATH ((problem))--
emitRandomThunkImpl calls jit.rshift64, which emits SAR (arithmetic shift).
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5311.
By contrast, urshift64 emits SHR (logical shift), this matches the intended
semantics.
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5350.

--IMPACT--
This discrepancy means that, when given the same seed, Math.random produces
different output depending on whether the JIT path is taken. The bug is not
immediately obvious because both arithmetic and logical right shifts still
generate values within the expected [0, 1) range, and the distribution may look
reasonable at a glance. However, the underlying state advancement diverges from
the intended algorithm, which leads to inconsistent and incorrect behavior
compared to other code paths.

* JSTests/stress/math-random-inline-vs-non-inline-paths.js:

Canonical link: <a href="https://commits.webkit.org/300785@main">https://commits.webkit.org/300785@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/212a76b3681a2297461fef6dfe7a8f412be3795e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130666 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52188 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126832 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/35305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74818 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28979 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74142 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116038 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29202 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133359 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122411 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50832 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51206 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107038 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26058 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47853 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50685 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56449 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153507 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39014 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on JSTests/stress/math-random-inline-vs-non-inline-paths.js:
```diff
@@ -0,0 +1,24 @@
+//@ runDefault("--useJIT=1")
```
If this test case reproduces without any extra flags, this line should be removed. Confirm that the test fails without your patch, and succeeds with it:

```
run-javascriptcore-tests --debug --no-build --jsc-only --no-fail-fast --memory-limited --jsc-stress --filter "random-inline"
```

matthewoestreich replied:
After making the suggested change, and using the command you provided:

## Results Without Patch

```bash
** The following JSC stress test failures have been introduced:
	stress/math-random-inline-vs-non-inline-paths.js.bytecode-cache
	stress/math-random-inline-vs-non-inline-paths.js.default
	stress/math-random-inline-vs-non-inline-paths.js.dfg-eager
	stress/math-random-inline-vs-non-inline-paths.js.dfg-eager-no-cjit-validate
	stress/math-random-inline-vs-non-inline-paths.js.eager-jettison-no-cjit
	stress/math-random-inline-vs-non-inline-paths.js.ftl-eager
	stress/math-random-inline-vs-non-inline-paths.js.ftl-eager-no-cjit-b3o1
	stress/math-random-inline-vs-non-inline-paths.js.ftl-no-cjit-b3o0
	stress/math-random-inline-vs-non-inline-paths.js.ftl-no-cjit-no-inline-validate
	stress/math-random-inline-vs-non-inline-paths.js.ftl-no-cjit-no-put-stack-validate
	stress/math-random-inline-vs-non-inline-paths.js.ftl-no-cjit-small-pool
	stress/math-random-inline-vs-non-inline-paths.js.ftl-no-cjit-validate-sampling-profiler
	stress/math-random-inline-vs-non-inline-paths.js.mini-mode
	stress/math-random-inline-vs-non-inline-paths.js.no-cjit-validate-phases
	stress/math-random-inline-vs-non-inline-paths.js.no-ftl
	stress/math-random-inline-vs-non-inline-paths.js.no-llint

Results for JSC stress tests:
    16 failures found.
    0 tests flaky but passed.
    0 tests flaky and failed.
    0 tests failed to complete.
```

## Results Patched

```bash
Results for JSC stress tests:
    0 failures found.
    0 tests flaky but passed.
    0 tests flaky and failed.
    0 tests failed to complete.
    OK.
```

matthewoestreich committed:
[JSC] Math.random JIT path uses arithmetic instead of logical right shifts
https://bugs.webkit.org/show_bug.cgi?id=299242

Reviewed by Justin Michaud.

--SUMMARY--
This report concerns Math.random output when taking the JIT path.
Currently, state advancement in emitRandomThunkImpl  uses arithmetic
right shifts (rshift64), whereas all other advancement paths use logical
right shifts.
[Source]: Source/JavaScriptCore/jit/AssemblyHelpers.cpp, line 847.

--REFERENCE IMPLEMENTATION--
WeakRandom::nextState uses the >> operator on a uint64_t, which performs
a logical shift.
[Source]: Source/WTF/wtf/WeakRandom.h, line 109.

--FTL LOWERING PATH--
compileArithRandom uses lShr (logical shift right).
[Source]: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp, line 3501.

--JIT PATH ((problem))--
emitRandomThunkImpl calls jit.rshift64, which emits SAR (arithmetic shift).
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5311.
By contrast, urshift64 emits SHR (logical shift), this matches the intended
semantics.
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5350.

--IMPACT--
This discrepancy means that, when given the same seed, Math.random produces
different output depending on whether the JIT path is taken. The bug is not
immediately obvious because both arithmetic and logical right shifts still
generate values within the expected [0, 1) range, and the distribution may look
reasonable at a glance. However, the underlying state advancement diverges from
the intended algorithm, which leads to inconsistent and incorrect behavior
compared to other code paths.

Test: JSTests/stress/math-random-inline-vs-non-inline-paths.js

Test: JSTests/stress/math-random-inline-vs-non-inline-paths.js
* JSTests/stress/math-random-inline-vs-non-inline-paths.js: Added.
(isEqualArr):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::emitRandomThunkImpl):

[JSC] Math.random JIT path uses arithmetic instead of logical right shifts
https://bugs.webkit.org/show_bug.cgi?id=299242

Reviewed by Justin Michaud.

--SUMMARY--
This report concerns Math.random output when taking the JIT path.
Currently, state advancement in emitRandomThunkImpl  uses arithmetic
right shifts (rshift64), whereas all other advancement paths use logical
right shifts.
[Source]: Source/JavaScriptCore/jit/AssemblyHelpers.cpp, line 847.

--REFERENCE IMPLEMENTATION--
WeakRandom::nextState uses the >> operator on a uint64_t, which performs
a logical shift.
[Source]: Source/WTF/wtf/WeakRandom.h, line 109.

--FTL LOWERING PATH--
compileArithRandom uses lShr (logical shift right).
[Source]: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp, line 3501.

--JIT PATH ((problem))--
emitRandomThunkImpl calls jit.rshift64, which emits SAR (arithmetic shift).
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5311.
By contrast, urshift64 emits SHR (logical shift), this matches the intended
semantics.
[Source]: Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h, line 5350.

--IMPACT--
This discrepancy means that, when given the same seed, Math.random produces
different output depending on whether the JIT path is taken. The bug is not
immediately obvious because both arithmetic and logical right shifts still
generate values within the expected [0, 1) range, and the distribution may look
reasonable at a glance. However, the underlying state advancement diverges from
the intended algorithm, which leads to inconsistent and incorrect behavior
compared to other code paths.

* JSTests/stress/math-random-inline-vs-non-inline-paths.js:

Canonical link: https://commits.webkit.org/300785@main
